### PR TITLE
MEPTS 255: Tx-ML Updates

### PR DESCRIPTION
**Pull request by ynurmahomed**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java [hunk 1]
```diff
@@ -68,4 +68,179 @@ public static String getTransferredOutPatients(int program, int state) {
...
+            + returnVisitDateForDrugsConcept
+            + " , "
+            + returnVisitDateConcept
+            + " ) "
+            + " AND e.location_id =:location "
+            + " GROUP BY p.patient_id "
+            + " )la ON la.patient_id=pa.patient_id "
+            + " WHERE (o.obs_datetime BETWEEN la.return_date AND :endDate) "
+            + " AND e.encounter_type NOT IN ( "
+            + homeVisitCardEncounterTypeId
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java]: I think it is missing the other encounter types mentioned in the spec (encounter type id = 21 or 36 or 37)

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java [hunk 2]
```diff
@@ -68,4 +68,179 @@ public static String getTransferredOutPatients(int program, int state) {
...
+            + " ) "
+            + " AND o.concept_id IN ( "
+            + returnVisitDateForDrugsConcept
+            + " , "
+            + returnVisitDateConcept
+            + " ) "
+            + " AND e.location_id =:location "
+            + " GROUP BY p.patient_id "
+            + " )la ON la.patient_id=pa.patient_id "
+            + " WHERE (o.obs_datetime BETWEEN la.return_date AND :endDate) "
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java]: Needs to be clarified in the spec where it says "reporting end date and the reporting end date"

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxMlCohortQueries.java [hunk 3]
```diff
@@ -105,10 +106,33 @@ public CohortDefinition getPatientsWhoMissedNextAppointmentAndNotTransferredOut(
...
+    sqlCohortDefinition.setName("Get patients marked as dead in Patient Home Visit Card");
+    sqlCohortDefinition.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    sqlCohortDefinition.addParameter(new Parameter("endDate", "End Date", Date.class));
+    sqlCohortDefinition.addParameter(new Parameter("location", "Location", Location.class));
+
+    sqlCohortDefinition.setQuery(
+        TxMlQueries.getPatientsMarkedDeadInHomeVisitCard(
+            hivMetadata.getBuscaActivaEncounterType().getEncounterTypeId(),
+            hivMetadata.getReasonPatientNotFound().getConceptId(),
+            hivMetadata.getPatientIsDead().getConceptId()));
```
[the remark this answers is not in the document]
Comment (KelvinWanyama) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxMlCohortQueries.java]: The concept Id was wrong. 1383 is for GOOD concept and we needed dead concept (2027), we fixed that with Pinki

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxMlCohortQueries.java [hunk 4]
```diff
@@ -129,4 +155,190 @@ public CohortDefinition getPatientsWhoMissedNextAppointmentAndNotTransferredOut(
...
+    cd.addSearch(
+        "missedAppointmentLessTransfers",
+        EptsReportUtils.map(
+            getPatientsWhoMissedNextAppointmentAndNotTransferredOut(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+    cd.addSearch(
+        "withoutVisitCard",
+        EptsReportUtils.map(
+            getAllPatientsWithoutVisitCard(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxMlCohortQueries.java]: @KelvinWanyama Shouldn't the date boundaries for `All patients without patient visit card` be `the last scheduled appointment or drugs pick date` and `reporting end date`?
I don't see how its limiting by last scheduled appointment.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxMlCohortQueries.java [hunk 5]
```diff
@@ -129,4 +155,190 @@ public CohortDefinition getPatientsWhoMissedNextAppointmentAndNotTransferredOut(
...
+            getPatientsWithoutVisitCardAndWithObs(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+    cd.addSearch(
+        "scheduledLastAppointment",
+        EptsReportUtils.map(
+            getPatientsLastScheduledAppointmentOrDrugPickup(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+
+    cd.setCompositionString(
+        "missedAppointmentLessTransfers AND ((withoutVisitCard AND scheduledLastAppointment) OR (withoutVisitCard AND scheduledLastAppointment AND patientsWithObs)");
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxMlCohortQueries.java]: Does `withoutVisitCard AND scheduledLastAppointment` mean `patients with encounters that are not visit card type and have an appointment scheduled between startDate and endDate`?

I think the previous query was closer to what we want. Maybe we should fetch all patients that do have visit card appointments and exclude them from (a and b).

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java [hunk 6]
```diff
@@ -68,4 +65,165 @@ public static String getTransferredOutPatients(int program, int state) {
...
+   Untraced Patients Criteria 1 And Untraced Patients Criteria 2
+   All Patients without “Patient Visit Card” (Encounter type 21 or 36 or 37)
+  */
+  public static String getAllPatientsWithoutVisitCard(int homeVisitCardEncounterTypeId) {
+    String query =
+        " SELECT pa.patient_id FROM patient pa "
+            + " INNER JOIN encounter e ON e.patient_id = pa.patient_id "
+            + " INNER JOIN obs o ON o.encounter_id = e.encounter_id "
+            + " WHERE e.encounter_type NOT IN ( "
+            + homeVisitCardEncounterTypeId
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java]: What about encounter types 36 and 37?

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java [hunk 7]
```diff
@@ -68,4 +65,165 @@ public static String getTransferredOutPatients(int program, int state) {
...
+            + " ) max_frida "
+            + " INNER JOIN obs o on o.person_id=max_frida.patient_id "
+            + " WHERE max_frida.encounter_datetime=o.obs_datetime AND o.voided=0 AND o.concept_id IN ( "
+            + returnVisitDateForDrugsConcept
+            + " , "
+            + returnVisitDateConcept
+            + " ) "
+            + " AND o.location_id=:location AND DATEDIFF(:endDate,o.value_datetime)<=:endDate";
+
+    return query;
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java]: Please use String.format

Comment (KelvinWanyama) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java]: okay

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxMlCohortQueries.java (hunk 4 above)
Comment (KelvinWanyama) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxMlCohortQueries.java]: When you look at the composition string, I am ANDing them with the scheduled last appointments or drug pick up.
I wrote separate queries and then joined them through the composition string.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java (hunk 7 above)
Comment (KelvinWanyama) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxMlQueries.java]: done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxMlCohortQueries.java (hunk 5 above)
Comment (KelvinWanyama) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxMlCohortQueries.java]: Fixed

**Pull request merged**